addons/link_tracker/models/mail_render_mixin.py

```diff
@@ -34,26 +38,72 @@ def _shorten_links(self, html, link_tracker_vals, blacklist=None, base_url=None)
 
         :return: updated html
         """
+        if not html or is_html_empty(html):
+            return html
         base_url = base_url or self.env['ir.config_parameter'].sudo().get_param('web.base.url')
         short_schema = base_url + '/r/'
-        for match in set(re.findall(tools.HTML_TAG_URL_REGEX, html)):
-            long_url = match[1]
-            # Don't shorten already-shortened links
-            if long_url.startswith(short_schema):
-                continue
-            # Don't shorten urls present in blacklist (aka to skip list)
-            if blacklist and any(s in long_url for s in blacklist):
-                continue
-            label = (match[3] or '').strip()
 
-            create_vals = dict(link_tracker_vals, url=unescape(long_url), label=unescape(label))
-            link = self.env['link.tracker'].search_or_create(create_vals)
-            if link.short_url:
-                # `str` manipulation required to support replacing "&" characters, common in urls
-                new_href = match[0].replace(long_url, link.short_url)
-                html = html.replace(markupsafe.Markup(match[0]), markupsafe.Markup(new_href))
+        root = lxml.html.fromstring(html)
```

Comment:
Note that you may keep a ``tools/url_tracker.py`` file with the tool functions; there is a balance between imperative and class-based code :) (don't hate me) . I prefer small and clear tools to black box classes (that was my main point of the previous review).